Gems/ROS2/Code/Source/Camera/CameraSensor.cpp

```diff
@@ -240,4 +253,51 @@ namespace ROS2
         return "Color";
     };
 
+    CameraRGBDSensor::CameraRGBDSensor(const CameraSensorDescription& cameraSensorDescription)
+        : CameraColorSensor(cameraSensorDescription)
+    {
+    }
+
+    void CameraRGBDSensor::ReadBackDepth(
+        AZStd::function<void(const AZ::RPI::AttachmentReadback::ReadbackResult& result)> callback)
+    {
+            AZ::Render::FrameCaptureId captureId = AZ::Render::InvalidFrameCaptureId;
+            AZStd::vector<AZStd::string> passHierarchy{m_pipelineName,"DepthPrePass"};
+            AZ::Render::FrameCaptureRequestBus::BroadcastResult(
+                captureId,
+                &AZ::Render::FrameCaptureRequestBus::Events::CapturePassAttachmentWithCallback,
+                passHierarchy,
+                AZStd::string("DepthLinear"),
+                callback,
+                AZ::RPI::PassAttachmentReadbackOption::Output);
+    }
+
+    void CameraRGBDSensor::RequestMessagePublication(
+        AZStd::vector<std::shared_ptr<rclcpp::Publisher<sensor_msgs::msg::Image>>> publishers,
+        const AZ::Transform& cameraPose,
+        const std_msgs::msg::Header& header)
+    {
+        AZ_Assert(publishers.size()==2, "RequestMessagePublication for CameraRGBDSensor should be called with exactly two publishers");
+        const auto publisherDepth = publishers.back();
+        ReadBackDepth(
+            [header, publisherDepth](const AZ::RPI::AttachmentReadback::ReadbackResult& result)
+            {
+                if (result.m_state == AZ::RPI::AttachmentReadback::ReadbackState::Success)
+                {
+                    const AZ::RHI::ImageDescriptor& descriptor = result.m_imageDescriptor;
+                    const auto format = descriptor.m_format;
+                    AZ_Assert(Internal::FormatMappings.contains(format), "Unknown format in result %u", static_cast<uint32_t>(format));
+                    sensor_msgs::msg::Image message;
+                    message.encoding = Internal::FormatMappings.at(format);
+                    message.width = descriptor.m_size.m_width;
+                    message.height = descriptor.m_size.m_height;
+                    message.step = message.width * Internal::BitDepth.at(format);
+                    message.data = std::vector<uint8_t>(result.m_dataBuffer->data(), result.m_dataBuffer->data() + result.m_dataBuffer->size());
```

Comment:
This should be able to be simplified to 
`message.data = *result.m_dataBuffer;`

Reply:
We move data from AZStd::vector to std::vector. The simplification is, sadly, not feasible.